### PR TITLE
Dynamically update public nodes

### PR DESCRIPTION
**Pull request by robertkiel**
Changes:
- use EventEmitter to inform `hopr-connect` about new potential relay nodes
- distinguish between `initialNodes` (that are known when calling the constructor) and `publicNodes` (that are *learned* at runtime and consumed by using events)
- run `yarn install` only on cache misses

Breaking changes:
- ignore config `bootstrapServers` and use `initialNodes` / `publicNodes` instead

Not covered:
- implement dynamic remove, hence not completely implementing #207

### CODE REVIEW

File: .github/workflows/integration-testing.yaml [hunk 1]
```diff
@@ -29,22 +29,21 @@ jobs:
...
-        id: nodejs-cache
+      - uses: actions/cache@v2
+        id: yarn-cache
         with:
           path: |
             node_modules
             */*/node_modules
             ${{ steps.yarn-cache-dir-path.outputs.dir }}
-          key: ${{ runner.os }}-nodejs-${{ hashFiles('**/yarn.lock') }}
+          key: ${{ runner.os }}-yarn-${{ hashFiles('**/yarn.lock') }}
```
Comment (andreykiryushkin) [inline .github/workflows/integration-testing.yaml:39]: is that relevant to the task?

File: src/base/listener.spec.ts [hunk 2]
```diff
@@ -369,4 +388,112 @@ describe('check listening to sockets', function () {
...
+      msgReceived: Defer()
+    })
+
+    node.publicNodesEmitter.emit(
+      `publicNode`,
+      new Multiaddr(`/ip4/127.0.0.1/tcp/${relay.listener.getPort()}/p2p/${relay.peerId.toB58String()}`)
+    )
+
+    // Let events happen
+    await new Promise((resolve) => setTimeout(resolve))
```
Comment (andreykiryushkin) [inline src/base/listener.spec.ts:454]: somehow I don't like this pattern, can it be done better?

File: src/base/listener.ts [hunk 3]
```diff
@@ -129,17 +123,86 @@ class Listener extends EventEmitter implements InterfaceListener {
...
+    }
+
+    const abort = new AbortController()
+    const timeout = setTimeout(abort.abort.bind(abort), RELAY_CONTACT_TIMEOUT)
+
+    const result = await this.connectToRelay(ma, maPeerId, { signal: abort.signal })
+
+    clearTimeout(timeout)
+
+    if (result.latency < 0) {
```
Comment (andreykiryushkin) [inline src/base/listener.ts:195]: how can latency be negative and why?

File: src/index.ts [hunk 4]
```diff
@@ -142,26 +126,26 @@ class HoprConnect implements Transport {
...
     if (this.__noWebRTCUpgrade) {
       verbose(`DEBUG mode: no WebRTC upgrade`)
     }
 
-    verbose(
-      `Created ${this[Symbol.toStringTag]} stack (Stun: ${this.stunServers
-        ?.map((server: Multiaddr) => server.toString())
-        .join(',')})`
-    )
+    verbose(`Created ${this[Symbol.toStringTag]} stack`)
```
Comment (andreykiryushkin) [inline src/index.ts:146]: this line is cryptic

**Review by andreykiryushkin — COMMENTED** [review]
probably not super-useful review

File: src/base/listener.ts (hunk 3 above)
Comment (robertkiel) [reply src/base/listener.ts:197]: negative latency == timeout

File: .github/workflows/integration-testing.yaml (hunk 1 above)
Comment (robertkiel) [reply .github/workflows/integration-testing.yaml:39]: small bugfix because Github actions used different cache keys before

File: src/base/listener.spec.ts (hunk 2 above)
Comment (robertkiel) [reply src/base/listener.spec.ts]: fixed, now using proper events

File: src/index.ts (hunk 4 above)
Comment (robertkiel) [reply src/index.ts]: removed because no longer necessary

**Review by andreykiryushkin — APPROVED** [review]
